tensorflow/core/kernels/mkl_conv_ops.cc

```diff
@@ -42,26 +43,288 @@ limitations under the License.
+    dst_fmt_ = static_cast<mkldnn::memory::format>(
+                 fwd_pd_.get()->dst_primitive_desc().desc().data.format);
+
+    // create memory primitive based on dummy data
+    src_mem_.reset(new memory(fwd_pd_.get()->src_primitive_desc(), DummyData));
+    filter_mem_.reset(new memory(fwd_pd_.get()->weights_primitive_desc(),
+                      DummyData));
+    dst_mem_.reset(new memory(fwd_pd_.get()->dst_primitive_desc(), DummyData));
+
+    // create convolution primitive and add it to net
+    if (!bias_dims.empty()) {
+        bias_mem_.reset(new memory({{{bias_dims}, MklDnnType<T>(),
+                                   memory::format::x}, cpu_engine_}, DummyData));
+        conv_fwd_.reset(new convolution_forward(*fwd_pd_, *src_mem_,
+                        *filter_mem_, *bias_mem_, *dst_mem_));
+    } else {
+        conv_fwd_.reset(new convolution_forward(*fwd_pd_, *src_mem_,
+                        *filter_mem_, *dst_mem_));
+    }
+
+    fwd_primitives_.push_back(*conv_fwd_);
+    return;
+  }
+
+ public:
+  // Convolution forward execute with bias
+  void Execute(void* src, void* w, void* b, void* dst) {
+    src_mem_->set_data_handle(src);
+    filter_mem_->set_data_handle(w);
+    bias_mem_->set_data_handle(b);
+    dst_mem_->set_data_handle(dst);
+    fwd_stream_->submit(fwd_primitives_);
+
+    // after exec, set data handle back
+    src_mem_->set_data_handle(DummyData);
+    filter_mem_->set_data_handle(DummyData);
+    bias_mem_->set_data_handle(DummyData);
+    dst_mem_->set_data_handle(DummyData);
+
+    return;
+  }
+
+  // Convolution forward execute without bias
+  void Execute(void* src, void* w, void* dst) {
+    src_mem_->set_data_handle(src);
+    filter_mem_->set_data_handle(w);
+    dst_mem_->set_data_handle(dst);
+    fwd_stream_->submit(fwd_primitives_);
+
+    // after exec, set data handle back
+    src_mem_->set_data_handle(DummyData);
+    filter_mem_->set_data_handle(DummyData);
+    dst_mem_->set_data_handle(DummyData);
+
+    return;
+  }
+
+  // expected memory format for this primitive instance
+  memory::format src_fmt_;
+  memory::format filter_fmt_;
+  memory::format dst_fmt_;
+
+  // convolution primitive
+  std::shared_ptr<mkldnn::convolution_forward::primitive_desc> fwd_pd_;
+  std::shared_ptr<mkldnn::primitive> conv_fwd_;
+
+ private:
+  // MKLDNN memory
+  std::shared_ptr<mkldnn::memory> src_mem_;
+  std::shared_ptr<mkldnn::memory> filter_mem_;
+  std::shared_ptr<mkldnn::memory> bias_mem_;
+  std::shared_ptr<mkldnn::memory> dst_mem_;
+
+  std::shared_ptr<mkldnn::stream> fwd_stream_;
+  std::vector<mkldnn::primitive> fwd_primitives_;
+
+  // desc & prmitive desc
+  std::shared_ptr<mkldnn::convolution_forward::desc> fwd_desc_;
+
+  // memory dims
+  mkldnn::memory::dims dilates_;
+  mkldnn::memory::dims strides_;
+  mkldnn::memory::dims padding_l_;
+  mkldnn::memory::dims padding_r_;
+
+  // memory desc
+  std::shared_ptr<mkldnn::memory::desc> src_md_;
```

Comment:
Similar comment as above. Is it worth organizing these memory descriptors in their own struct?

Reply:
Hi Andy, for now let's keep these memory descriptors as they now. 
I will follow up by discussing your suggestion with our team members
and refactor (if necessary) in new future PRs (primitive reuse for conv2d bwd,
batchnorm, relu, pooling ops etc.). Thanks!  GZ